### ISSUE TITLE
Auto Updating doesn't appear to be working for me

### ISSUE THREAD

**Issue by sachabarber** [Squirrel/Squirrel.Windows#365]
I am new to Squirrel for Windows, but have some exposure to ClickOnce.

<strong>General Notes</strong>
I am using the standard "Releases" folder to host my releases. I have a working web site which looks at a "dist" folder.

<strong>Initial Release</strong>
So after a bit of pain I have something almost working here.
1. I create the nuspec file
2. I create the nuget package
3. I do a release of current app version
4. Copy from releases to dist folder
5. run setup.exe from browser

App is install all is good.

<strong>New Release</strong>
1. I then edit my codebase to contain the following
   
   ```
       var updateManager = new UpdateManager(@"http://localhost:85/", "SqTest2");
       using (updateManager)
       {
           var updateInfo = await updateManager.CheckForUpdate();
           if (updateInfo == null || !updateInfo.ReleasesToApply.Any())
           {
               return;
           }
           else
           {
               var releases = updateInfo.ReleasesToApply;
               await updateManager.DownloadReleases(releases, _ => { });
               await updateManager.ApplyReleases(updateInfo, _ => { });
           }
       } 
   ```
2. I then also change the assembly version
3. I create the nuget package (this generates full/delta nuget packages)
4. I do a release of current app version
5. Copy from releases to dist folder

At this point I am expecting to be able to run installed app, and it auto update to latest version. This does NOT happen, it just runs old version. 

Strange thing is, if I just run the web site setup,exe again I do get the latest version.

**Comment by paulcbetts** [comment]
The HTTP URL is obscuring what's going on a bit, can you just use a path instead to illustrate where you're pulling updates from?

**Comment by sachabarber** [comment]
So I have a web site which is http://localhost (port 85) which is set up to point to the folder C:\Users\barbers\Desktop\SqTest2\dist. The new releases are built to C:\Users\barbers\Desktop\SqTest2\Releases which for me right now contains the following list of files

RELEASES
Setup.exe
SqTest2-1.0.0.0-full.nupkg
SqTest2-1.1.0.0-delta.nupkg
SqTest2-1.1.0.0-full.nupkg
SqTest2-1.2.0.0-delta.nupkg
SqTest2-1.2.0.0-full.nupkg
SqTest2-1.3.0.0-delta.nupkg
SqTest2-1.3.0.0-full.nupkg
SqTest2-1.4.0.0-delta.nupkg
SqTest2-1.4.0.0-full.nupkg

I then copy all of these to the "dist" folder C:\Users\barbers\Desktop\SqTest2\dist (which acts as the folder for the IIS web site).


Do you need more information than that? Are you saying the autoupdater stuff in the app should be pointing to an actual path? That doesn't make too much sense to me, I would think this should be a url of well known web site.

Like I say though I a completely new to Squirrel for Windows, and just want to get it working as I loath Mage.exe (lost more hours to that EXE than I care to mention)

**Comment by paulcbetts** [comment]
> Are you saying the autoupdater stuff in the app should be pointing to an actual path? That doesn't make too much sense to me, I would think this should be a url of well known web site.

In production, the autoupdater stuff will point to an HTTPS URL, just like you've set it up. However, when you're testing updates locally, you can instead just point it to a local folder, because that's way easier to set up / debug, than spinning up a dummy HTTP server

**Comment by paulcbetts** [comment]
What happens to your local installed app folder when you run the code (i.e. `C:\Users\NAME\AppData\Local\SqTest2`? Do you see a new `app-VERSION` folder?)

**Comment by sachabarber** [comment]
Yes I see a folder called this

<strong>C:\Users\barbers\AppData\Local\SqTest2</strong>

Which contains the following images

<ul>
<li>.dead</li>
<li>app-1.3.0.0</li>
<li>app-1.4.0.0</li>
<li>packages</li>
<li>SquirrelSetup.log</li>
<li>Update.exe</li>
</ul>

**Comment by paulcbetts** [comment]
@sachabarber Ok - what does your app's shortcut point to?

**Comment by sachabarber** [comment]
I am not at work now (where I am trying this all out). I suspect I just need to change the auto updater to use a folder. The web site I setup is not HTTPS, it is standard HTTP.

I'll try and get it working with directory auto updater, and if that works I plan to write this all up in a semi-useful guide.

It looks very promising, but could be documented a bit better (sorry dude). It is good work no doubt there, just a bit hard to piece the bits together so far.

**Comment by sachabarber** [comment]
Paul

So I check out where the app shortcut was pointing to. It points to the folder

<strong>C:\Users\barbers\AppData\Local\SqTest2</strong>




So I tried using that for the UpdateManager path instead of web Url. So I then made a new release, and copied the auto generated release folder items to the path 

<strong>C:\Users\barbers\Desktop\SqTest2\dist</strong>

I then ran the setup.exe from that path. This installed everything fine. I then tried to run the installed app. When I start the app, it crashed saying the folder <strong>C:\Users\barbers\AppData\Local\SqTest2</strong> did not contain a directory for the current version.  At this point the installed app is configured to have its update path set to <strong>C:\Users\barbers\AppData\Local\SqTest2</strong>, and the shortcut also points to this folder.


I then checked the <strong>C:\Users\barbers\AppData\Local\SqTest2</strong>, and it <strong>does not</strong> contains a folder for the current release.

So based on that, my assumption was that this <strong>C:\Users\barbers\AppData\Local\SqTest2</strong>, can't be the one that is used for the UpdateManager path, as it only seems to get a new version of the app copied there after a successful run. Which seems very strange as this is the folder with the update.exe in it.

So I uninstalled the app.

Then created a new release where I changed the UpdateManager to be the <strong>C:\Users\barbers\Desktop\SqTest2\dist</strong> folder. Did a new release, copied the files to <strong>C:\Users\barbers\Desktop\SqTest2\dist</strong>, ran setup.exe from <strong>C:\Users\barbers\Desktop\SqTest2\dist</strong>. This installed a new app which is runnable, and also ends up creating a new version of the app in the <strong>C:\Users\barbers\AppData\Local\SqTest2</strong>. 

I then thought ok it did not crash, perhaps all I need to do now is create yet another new release and copy the files to the <strong>C:\Users\barbers\Desktop\SqTest2\dist</strong> (which is where setup.exe was run from), and it will auto update.

It didn't. Which does not suprise me, as the app shortcut is setup to point to the <strong>C:\Users\barbers\AppData\Local\SqTest2</strong>, which is also where update.exe is.


What does confuse the buggery out of me, is how I a meant to configure this UpdateManager path to point to the <strong>C:\Users\barbers\AppData\Local\SqTest2</strong> directory when this seems to crash as soon as I do a new release, due to the required app-x.x.x.x not being present there.

**Comment by stajs** [comment]
To risk stating the obvious, it looks like you have a pathing problem: you aren't telling the `UpdateManager` the right path to your `RELEASES` file. I found it a bit hard to follow your previous comment, but it looks like you just need to tell the `UpdateManager` to use the `dist` folder path. And I'd stick with folder path over a URL until you get it working locally.

If it helps to see a working example (and I'm not saying this is best practice), I half-assed-automated the packaging of our app and documented the release process here:

https://github.com/lic-nz/Malone/wiki/Release

I made the update URL a config setting, so you can modify the installed versions config to test against a local path. HTH

**Comment by sachabarber** [comment]
I agree this post of mine is hard to follow, but so is the NON existent Squirel documentation :-)

I could not see in your example where you even use the suggested "updateManager" code. I watched a video of Paul Betts showing Scott Hansleman how to use Squirrel. And even in that video he put the code in for UpdateManager, and then went on to comment it out before showing it working.


I basically have it all working apart from the Auto update stuff. I need to try that again. I like how you are creating the Nuget in code though they seems good.

**Comment by stajs** [comment]
For reference, this is where I check for updates:

https://github.com/lic-nz/Malone/blob/master/LIC.Malone.Client.Desktop/ViewModels/AppViewModel.cs#L444

I use a slightly expanded version of that "suggested" code from your initial comment. You'll see that the method is marked `async` in order to allow the `await`'s. Without seeing what modifiers you have on your containing method, it's possible adding the `Task.Run()` ended up being another way to do the same thing.

As for your update, I'm not sure what is happening there. With Squirrel built on nuget concepts you may be right that having a `packages.config` just kills it. But why do you need one? All your dependencies should be explicitly included...? Anyway, I'd suggest creating a separate issue to avoid mixing these bug threads and also properly formatting that stack trace. ATM, I can't tell if GitHubs auto-formatting has decided to truncate or mangle some important information in there.

**Comment by sachabarber** [comment]
I am not sure what you mean about why would I need a package.config. I do this

1. I have a UI which uses several NuGet packages, so there is a package.config that contains all the packages used by the UI in the project
2. I then create a NuSpec file for the UI project, where I create the NuSpec in same place as UIs csproj file location (which happens to end up in the same folder as the actual UIs  package.config file)
3. I then realeasify everything, and that breaks due to (apparently) having a package.config in the same folder as the nuspec file (I guess). Perhaps I should create the Nuspec file to new folder and do realeasify from a folder that does not have a package.config in it


I agree I should move some of this to new thread though. Ill try that, thanks for trying to help me here

**Comment by stajs** [comment]
Sorry, I think I misunderstood you. I thought you might have been including a `packages.config` inside your `nupkg`.

What version of Squirrel are you on? I'm using 0.99.0 and can't repro that issue. A few things:

0. Yeah, I build out to a separate location, but I verified that I could still build all back in the root of my UI's `csproj`. And my layout is the same as yours: `csproj`, `packages.config` and `nuspec` all in the same folder.
![image](https://cloud.githubusercontent.com/assets/2253814/8390345/544ed9c4-1ce3-11e5-998d-750be36675e1.png)

0. I also tested adding a `packages.config` in the folder where I build, and it worked fine:
![image](https://cloud.githubusercontent.com/assets/2253814/8390348/8a987508-1ce3-11e5-8689-e19095505ef4.png)

0. You missed describing a step above where you create a suitable `nupkg`. I assume that is just an oversight and you actually do have one. And by "suitable" I mean the dependencies are explicitly included, you aren't doing anything weird with recursively nuget'ing things... I dunno just chucking ideas out there. It might be worth showing the contents of your `nupkg` to check.

0. I call Squirrel from the command line, but also tested using the Package Manager Console inside VS.

The releasify process AFAIK will only operate on that single `nupkg`, so I don't see how having a `packages.config` sitting next to it matters. And it doesn't make a difference from my testing. It could be worth you forking my project and try releasifying that, because if it that doesn't work for you it might point to something being borked with your environment.

**Comment by Thieum** [comment]
@shiftkey discussion has ended with no further issues - this can be closed.

**Comment by OfficialMudasir** [comment]
> The HTTP URL is obscuring what's going on a bit, can you just use a path instead to illustrate where you're pulling updates from?

can you please tell me how to generate Update.exe after issuing the command squirrel --releasify "updated nuget package name".nupkg because I am not able to generate that.

**Comment by vernou** [comment]
The command `Squirrel --releasify` generate "Setup.exe". This executable install the app and deploy the "Update.exe" in `%localappdata%\<AppName>\Update.exe`.

But the "Update.exe" is "Squirrel.exe" (used in the command `Squirrel --releasify`) that is duplicated, renamed and integrated in the "Setup.exe".  You can :

1. Download the packge NuGet [Squirrel.Windows](https://www.nuget.org/packages/squirrel.windows)
2. Unzip the package (nuspkg is just a zip)
3. Go in "tools" folder
4. Copy the file "Squirrel.exe"
5. Rename the copy  "Update.exe"
